src/nvim/lua/executor.c

```diff
@@ -78,7 +78,22 @@ static void nlua_error(lua_State *const lstate, const char *const msg)
   FUNC_ATTR_NONNULL_ALL
 {
   size_t len;
-  const char *const str = lua_tolstring(lstate, -1, &len);
+  const char *str = NULL;
+
+  if (luaL_getmetafield(lstate, -1, "__tostring")) {
+    if (lua_isfunction(lstate, -1) && luaL_callmeta(lstate, -2, "__tostring")) {
+      // has __tostring and is a function, call it, convert the result and pop.
+      str = lua_tolstring(lstate, -1, &len);
+      lua_pop(lstate, 1);
+    }
```

Comment:
Possible to add an `else` here along the lines of "you gave a `__tostring` but it wasnt callable" if desired, right now we just  revert to the current behaviour of outputting `[NULL]` (via `lua_tolstring(<some-table>)`).